jax/experimental/jax2tf/jax2tf.py

```diff
@@ -1120,31 +1120,41 @@ def _scatter_shape(operand, scatter_indices, updates, dimension_numbers):
       dimension_numbers=dimension_numbers)
   return out.shape
 
+def _mk_typed_jaxpr(jaxpr: core.Jaxpr, literals: Sequence) -> core.TypedJaxpr:
+  return core.TypedJaxpr(jaxpr, literals,
+                         tuple(map(lambda v: v.aval, jaxpr.invars)),
+                         tuple(map(lambda v: v.aval, jaxpr.outvars)))
+
+@functools.partial(bool_to_int8, argnums=(0, 2))
+def _scatter(operand, scatter_indices, updates, update_jaxpr, update_consts,
+             dimension_numbers, indices_are_sorted, unique_indices):
+  del unique_indices
+  assert len(update_consts) == 0, "Update computation cannot have constants"
 
-@functools.partial(bool_to_int8, argnums=(1, 3))
-def _scatter(update_computation, operand, scatter_indices, updates,
-             update_jaxpr, update_consts, dimension_numbers,
-             indices_are_sorted, unique_indices):
-  """Tensorflow implementation of scatter with an update computation."""
-  del update_jaxpr, update_consts, unique_indices
   out_shape = _scatter_shape(operand, scatter_indices, updates,
                              dimension_numbers)
   proto = _scatter_dimensions_proto(scatter_indices.shape, dimension_numbers)
-  o_spec = tf.TensorSpec(None, dtype=operand.dtype)
+
+  def update_computation(arg1: TfVal, arg2: TfVal) -> TfVal:
+    typed_jaxpr = _mk_typed_jaxpr(update_jaxpr, update_consts)
+    res, = _interpret_jaxpr(typed_jaxpr, arg1, arg2)
+    return res
+
+  o_spec = tf.TensorSpec((), dtype=operand.dtype)
   xla_update_computation = (
-      tf.function(update_computation).get_concrete_function(o_spec, o_spec))
-  # We compile due to TF bug, see comment on gather
-  out = tf.function(lambda o, s, u: tfxla.scatter(o, s, u, xla_update_computation, proto,
-                                                  indices_are_sorted=indices_are_sorted),
-                    experimental_compile=True)(operand, scatter_indices, updates)
+    tf.function(update_computation).get_concrete_function(o_spec, o_spec))
+
+  out = tfxla.scatter(operand, scatter_indices, updates, xla_update_computation,
```

Comment:
I would keep the `tf.function` wrapper for now, I suspect that this change will fail on GPU. I have PR #4030 to finish to clean this.

Reply:
Added back the wrapper in the latest commit.